### PR TITLE
Use current working directory if no `package.json`

### LINKED ISSUE #1041: TypeError: Path must be a string. Received null
## Description

ava nor ava --init nor ava --version is working.
The project has no config at all and no node_modules.
ava is installed globaly

### Error Message & Stack Trace

```
d:\repositories\git\crmintern\Testing_Ground>ava --version                                                                                                                             
path.js:7                                                                                                                                                                              
    throw new TypeError('Path must be a string. Received ' + inspect(path));                                                                                                           
    ^                                                                                                                                                                                  
                                                                                                                                                                                       
TypeError: Path must be a string. Received null                                                                                                                                        
    at assertPath (path.js:7:11)                                                                                                                                                       
    at Object.dirname (path.js:697:5)                                                                                                                                                  
    at Object.<anonymous> (C:\Users\stefan.schneider\AppData\Roaming\npm\node_modules\ava\cli.js:43:19)                                                                                
    at Module._compile (module.js:556:32)                                                                                                                                              
    at Object.Module._extensions..js (module.js:565:10)                                                                                                                                
    at Module.load (module.js:473:32)                                                                                                                                                  
    at tryModuleLoad (module.js:432:12)                                                                                                                                                
    at Function.Module._load (module.js:424:3)                                                                                                                                         
    at Module.runMain (module.js:590:10)                                                                                                                                               
    at run (bootstrap_node.js:394:7)                                                                                                                                                   
    at startup (bootstrap_node.js:149:9)                                                                                                                                               
    at bootstrap_node.js:509:3  
```

### Command-Line Arguments

Copy your npm build scripts or the `ava` command used:

```
ava --version
```

## Environment

Tell us which operating system you are using, as well as which versions of Node.js, npm, and AVA. Run the following to get it quickly:

```
node: v6.6.0
ava: v0.16.0
npm: 3.10.8
VSCode Version: Code 1.5.2 (66f37fd2a99eb9d628dd374d81d78835b410c39b, 2016-09-12T13:30:03.456Z)
OS Version: Windows_NT ia32 10.0.10586
```

### PR DESCRIPTION
**Pull request by sotojuan**
Fixes #1041.

Currently we give a nasty error when AVA is ran in a directory with no `package.json`. If none is found, we fallback to the current working directory.

### CODE REVIEW

**Comment by sotojuan** [comment]
This might interfere with https://github.com/avajs/ava/issues/32, not sure.

**Comment by sotojuan** [comment]
Closed by accident!

File: cli.js [hunk 1]
```diff
@@ -43,7 +43,13 @@ Promise.longStackTraces();
 
 var conf = pkgConf.sync('ava');
 
-var pkgDir = path.dirname(pkgConf.filepath(conf));
+var pkgDir;
+
+try {
+	pkgDir = path.dirname(pkgConf.filepath(conf));
+} catch (err) {
```
Comment (novemberborn) [inline cli.js]: Check if `pkgConf.filepath(conf)` returns `null` instead. If it does, use `process.cwd()`, else call `path.dirname()` with the return value.

**Comment by sotojuan** [comment]
Done!

File: cli.js [hunk 2]
```diff
@@ -43,7 +43,14 @@ Promise.longStackTraces();
...
 
-var pkgDir = path.dirname(pkgConf.filepath(conf));
+var pkgDir;
+var filepath = pkgConf.filepath(conf);
+
+if (filepath === null) {
+	pkgDir = process.cwd();
+} else {
+	pkgDir = path.dirname(filepath);
+}
```
Comment (sindresorhus) [inline cli.js]: Would be better as a simple ternary:

```js
var pkgDir = filepath === null ? process.cwd() : path.dirname(filepath);
```

**Comment by sotojuan** [comment]
Right, of course :p Fixed.

**Comment by jamestalmage** [comment]
Any way to test this?

I'm thinking something like:

```js
const dir = uniqueTempDir({create: true});
const testFilePath = path.join(dir, 'test.js');
const cliPath = require.resolve('../../cli.js');

fs.writeFileSync(testFilePath, `
  import test from 'ava';
  
  test(t => {
    t.pass();
  });
`);

execa(process.execPath, [cliPath], {cwd: dir}).then(endTheTest);
```

**Comment by sindresorhus** [comment]
@sotojuan Could you add a test?

**Comment by sotojuan** [comment]
Yep, sorry I got busy these past two weeks!

**Comment by sotojuan** [comment]
Any idea of why it's failing? Works locally (and test critique please).

**Comment by vdemedes** [comment]
Also passed locally on clean node_modules without a problem.

File: test/cli.js [hunk 3]
```diff
@@ -368,3 +371,17 @@ test('prefers local version of ava', function (t) {
 	t.ok(debugSpy.calledWith('Using local install of AVA'));
 	t.end();
 });
+
+test('use current working directory if `package.json` is not found', function (t) {
+	var dir = uniqueTempDir({create: true});
+	var testFilePath = path.join(dir, 'test.js');
+	var cliPath = require.resolve('../cli.js');
+	var avaPath = require.resolve('../index.js');
```
Comment (sindresorhus) [inline test/cli.js]:
```js
var avaPath = require.resolve('../');
```

File: test/cli.js [hunk 4]
```diff
@@ -368,3 +371,17 @@ test('prefers local version of ava', function (t) {
 	t.ok(debugSpy.calledWith('Using local install of AVA'));
 	t.end();
 });
+
+test('use current working directory if `package.json` is not found', function (t) {
+	var dir = uniqueTempDir({create: true});
```
Comment (sindresorhus) [inline test/cli.js]: `dir` => `cwd`

File: test/cli.js [hunk 5]
```diff
@@ -368,3 +371,17 @@ test('prefers local version of ava', function (t) {
...
+	var testFilePath = path.join(dir, 'test.js');
+	var cliPath = require.resolve('../cli.js');
+	var avaPath = require.resolve('../index.js');
+
+	fs.writeFileSync(testFilePath, 'import test from \'' + avaPath + '\';\ntest(t => { t.pass(); });');
+
+	execa(process.execPath, [cliPath], {cwd: dir}).then(() => {
+		t.pass();
+		t.end();
+	});
```
Comment (sindresorhus) [inline test/cli.js]: It needs to handle a possibly rejected promise too.

Comment (vdemedes) [inline test/cli.js]: I think you should just return the promise. You don't test the output of the command, so no need for manual then/catch. Return a promise and test will fail, if the promise rejects.

Comment (sindresorhus) [inline test/cli.js]: 👍 Ah true. Forgot node-tap got support for Promises.

**Comment by sotojuan** [comment]
Still breaking, but updated.

**Comment by novemberborn** [comment]
@sotojuan I've found why the test was failing, there's another bug where we assume we can always find a cache dir. That's not true when running outside of a directory with a `package.json` file!

Pushed a fix for that. Also merged master into this branch which is a bit unconventional but I didn't want to force-push ;-) Squash & merge will take care of that for us.

**Comment by sotojuan** [comment]
Phew, had to fix some git stuff and merge latest master. If the tests pass I can squash.

**Comment by sotojuan** [comment]
Boo AppVeyor.

**Comment by sindresorhus** [comment]
AppVeyor error:

> Error: Cannot find module 'C:projectsavaindex.js

My first thought; Where are the path separators?

File: test/cli.js [hunk 6]
```diff
@@ -368,3 +371,14 @@ test('prefers local version of ava', function (t) {
...
 	t.end();
 });
+
+test('use current working directory if `package.json` is not found', function () {
+	var cwd = uniqueTempDir({create: true});
+	var testFilePath = path.join(cwd, 'test.js');
+	var cliPath = require.resolve('../cli.js');
+	var avaPath = require.resolve('../');
+
+	fs.writeFileSync(testFilePath, 'import test from \'' + avaPath + '\';\ntest(t => { t.pass(); });');
```
Comment (novemberborn) [inline test/cli.js]: Changing to this probably fixes the AppVeyor error:

```js
'import test from ' + JSON.stringify(avaPath) + ';\ntest(t => { t.pass(); });'
```

I reckon on Windows the path contains backslashes which are interpreted as escape characters.

**Comment by sotojuan** [comment]
All right I'm pretty sure AppVeyor will like these changes.

**Comment by sotojuan** [comment]
🐔

**Comment by sotojuan** [comment]
Thanks!

**Pull request merged**